File: mns-reactor/src/test/java/cz/fb/manaus/reactor/betting/proposer/PriceProposalServiceTest.java
package cz.fb.manaus.reactor.betting.proposer;

import cz.fb.manaus.core.model.Side;
import cz.fb.manaus.core.test.AbstractLocalTestCase;
import cz.fb.manaus.reactor.betting.BetContext;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.OptionalDouble;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class PriceProposalServiceTest extends AbstractLocalTestCase {

    public static final List<PriceProposer> PROPOSERS = Arrays.asList(
            new TestProposer1(),
            new TestProposer2(),
            new FooProposer());
    public static final double MIN_PRICE = 1.5d;
    public static final double MAX_PRICE = 2d;
    @Autowired
    private PriceProposalService service;

    @Test
    public void testBackPrice() throws Exception {
        checkProposal(MAX_PRICE, Side.BACK, "testProposer2,fooProposer", PROPOSERS);
    }

    @Test
    public void testLayPrice() throws Exception {
        checkProposal(MIN_PRICE, Side.LAY, "testProposer1", PROPOSERS);
    }

    @Test(expected = IllegalStateException.class)
    public void testMandatoryPrice() throws Exception {
        PriceProposer proposer = ctx -> OptionalDouble.empty();
        service.reducePrices(mock(BetContext.class), Collections.singletonList(proposer), Side.LAY);
    }

    private void checkProposal(double expectedPrice, Side side, String proposer, List<PriceProposer> proposers) {
        BetContext context = mock(BetContext.class);
        when(context.getProperties()).thenReturn(new HashMap<>());
        double price = service.reducePrices(context, proposers, side).getPrice();
        assertThat(price, is(expectedPrice));
    }

    private static class TestProposer1 implements PriceProposer {

        @Override
        public OptionalDouble getProposedPrice(BetContext context) {
            return OptionalDouble.of(MIN_PRICE);
        }
    }

    private static class TestProposer2 implements PriceProposer {

        @Override
        public OptionalDouble getProposedPrice(BetContext context) {
            return OptionalDouble.of(MAX_PRICE);
        }
    }

    private static class FooProposer extends TestProposer2 {
    }


}

File: mns-reactor/src/test/java/cz/fb/manaus/reactor/betting/proposer/PriceProposalServiceTest.kt
package cz.fb.manaus.reactor.betting.proposer

import cz.fb.manaus.core.model.Side
import cz.fb.manaus.reactor.betting.BetEvent
import cz.fb.manaus.reactor.betting.HOME_EVENT_BACK
import cz.fb.manaus.reactor.betting.HOME_EVENT_LAY
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class PriceProposalServiceTest {

    private val service = PriceProposalService

    @Test
    fun `back price`() {
        checkProposal(MAX_PRICE, Side.BACK, PROPOSERS)
    }

    @Test
    fun `lay price`() {
        checkProposal(MIN_PRICE, Side.LAY, PROPOSERS)
    }

    @Test
    fun `mandatory price`() {
        val proposer: PriceProposer = object : PriceProposer {
            override fun getProposedPrice(event: BetEvent): Double? {
                return null
            }
        }
        assertFailsWith<IllegalStateException> {
            service.reducePrices(HOME_EVENT_LAY, listOf(proposer))
        }
    }

    private fun checkProposal(expectedPrice: Double, side: Side, proposers: List<PriceProposer>) {
        val event = when (side) {
            Side.LAY -> HOME_EVENT_LAY
            Side.BACK -> HOME_EVENT_BACK
        }
        val price = service.reducePrices(event, proposers).price
        assertEquals(expectedPrice, price)
    }

    private class TestProposer1 : PriceProposer {

        override fun getProposedPrice(event: BetEvent): Double {
            return MIN_PRICE
        }
    }

    private open class TestProposer2 : PriceProposer {

        override fun getProposedPrice(event: BetEvent): Double {
            return MAX_PRICE
        }
    }

    private class FooProposer : TestProposer2()

    companion object {

        val PROPOSERS = listOf(
            TestProposer1(),
            TestProposer2(),
            FooProposer()
        )
        const val MIN_PRICE = 1.5
        const val MAX_PRICE = 2.0
    }
}